recipe/common.php

```diff
@@ -46,41 +47,83 @@
-
-set('http_user', false);
-set('http_group', false);
+set('keep_releases', 5, function ($value) {
+    if (is_integer($value) && $value >= -1) {
+        return true;
+    }
+    throw new ConfigurationException("Config error: \"keep_releases\" value must be an integer equal or bigger than \"-1\".");
+});
 
-set('clear_paths', []);         // Relative path from release_path
-set('clear_use_sudo', false);    // Using sudo in clean commands?
+set('repository', '', function ($value) {
+    return is_string($value); 
+}); // Repository to deploy.
+
+set('shared_dirs', [], 'is_array');
+set('shared_files', [], 'is_array');
+set('copy_dirs', [], 'is_array');
+set('writable_dirs', [], 'is_array');
+set('writable_mode', 'acl', function ($value) {
+    return $value === 'chmod'
+        || $value === 'chown'
+        || $value === 'chgrp'
+        || $value === 'acl'; 
+}); // chmod, chown, chgrp or acl.
+
+set('writable_use_sudo', false, 'is_bool'); // Using sudo in writable commands?
+set('writable_recursive', true, 'is_bool'); // Common for all modes
+
+set('writable_chmod_mode', '0755', function ($value) {
+    $chmodPattern = '/^[ugoa]*([-+=]([rwxXst]*|[ugo]))+|[-+=]?[0-7]+$/';
+    return preg_match($chmodPattern, $value);
+}); // For chmod mode
+
+set('writable_chmod_recursive', true, 'is_bool'); // For chmod mode only (if is boolean, it has priority over `writable_recursive`)
+set('http_user', false, 'is_bool');
+set('http_group', false, 'is_bool');
+
+set('clear_paths', [], function ($value) {
+    if (!is_array($value)) {
+        return false;
+    }
+    foreach ($value as $dir) {
+        if (!is_dir($dir)) {
+            throw new ConfigurationException("Config error: folder \"$dir\" not found in \"clear_paths\" option.");
+        }
+    }
+    return true;
+});         // Relative path from release_path
 
-set('cleanup_use_sudo', false); // Using sudo in cleanup commands?
+set('clear_use_sudo', false, 'is_bool');    // Using sudo in clean commands?
+set('cleanup_use_sudo', false, 'is_bool'); // Using sudo in cleanup commands?
 
 set('use_relative_symlink', function () {
     return commandSupportsOption('ln', '--relative');
 });
+
 set('use_atomic_symlink', function () {
     return commandSupportsOption('mv', '--no-target-directory');
 });
 
-set('composer_action', 'install');
-set('composer_options', '{{composer_action}} --verbose --prefer-dist --no-progress --no-interaction --no-dev --optimize-autoloader --no-suggest');
+set('composer_action', 'install', function ($value) {
+    switch ($value) {
+    case 'install':
+    case 'i':
+    case 'require':
+    case 'u':
+    case 'update':
+    case 'upgrade':
+        return true;
+    }
+    return false;
+});
+
+set('composer_options', '{{composer_action}} --verbose --prefer-dist --no-progress --no-interaction --no-dev --optimize-autoloader --no-suggest', function ($value) {
+    return is_string($value);
```

Comment:
Use string, please.

Reply:
Here as well!

Reply:
I noticed you removed the `composer_options` and a few others, so I am removing them as well.